### DISCUSSION TITLE
Errors when using tower sessions with sqlite store

### DISCUSSION THREAD

**Discussion by CriusDe** [maxcountryman/tower-sessions#183] [Q&A]
Hi,

we get some errors in our application after the switch from axum sessions to tower sessions. 

    let pool = SqlitePool::connect(":memory:").await.unwrap();

    let sqlite_store = SqliteStore::new(pool);
    sqlite_store.migrate().await.unwrap();
    tokio::task::spawn(
        sqlite_store
            .clone()
            .continuously_delete_expired(tokio::time::Duration::from_secs(10)),
    );

    let session_layer = SessionManagerLayer::new(sqlite_store)
        .with_secure(true)
        .with_expiry(Expiry::OnInactivity(Duration::hours(2)));

        
   This is our init of the session store that is called once on server startup. 
   
   failed to save session err=error returned from database: (code: 1) no such table: tower_sessions
   error=error returned from database: (code: 1) no such table: tower_sessions
   
   these are our error logs we receive all the time. 
   
   Does anybody have an Idea or hint what the problem may be?

**Comment by maxcountryman** [comment]
Does the SQLite example work for you independent of your application?

Regarding your application, if the migration is succeeding then it's strange you'd see such an error. Are you sure the migration is working as expected?

**Reply by CriusDe** [reply]
I have not tested the example in isolation on our environment. So I can't really say if it is running independent of the application.

Hmm depends on what you mean with working as expected. We had to make a change in storage, because we used a cookiestore before and now switched to the in memory sqlite store. As far as functionality goes, everything works, beside from some random session terminations, but that could also happen on the client side. I can retrieve all sessions and work with them. The only thing that is strange are these errors. They also come in intervals of about half an hour for around 1-2 minutes (~200 errors) I can't really align them with any events happening on the server. So no correlation to traffic e.g.

I also tested to set the expiry to 4 hours and the continously delete task to 60 seconds and have no errors so far in the logs.

**Reply by maxcountryman** [reply]
> now switched to the in memory sqlite store

Have you considered using MokaStore instead? The primary difference is the LRU cache is bounded (altho Moka might allow you to configure an unbounded cache if you wanted) and of course the fact it's intended specifically for in-memory use.

**Reply by CriusDe** [reply]
Yeah i thought about that, i will give it try. Do you recommend using just the moka store or with sqlite as a backend like in the examples?

**Reply by maxcountryman** [reply]
I would recommend using `MokaStore` directly, at least at first. If that's working as expected, then you could use it with `CachingStore` and some other store (usually this second store would be persistent).

If using SQLite, in-memory might not be the best choice for a production setting. That's why the `axum-login` examples use a persistent file path.

**Reply by maxcountryman** [reply]
Regarding SQLite, in case it's something you do want to explore more, there's also the Rusqlite store: https://github.com/patte/tower-sessions-rusqlite-store. It's possible it could perform more reliably for your use case.

**Reply by CriusDe** [reply]
Ok thanks for your input, i will try and have a look at these solutions, to see what performs best
